### PR TITLE
[13.0] Add stock_checkout_sync

### PR DESCRIPTION
**Pull request by guewen**
Replaces #862 

Regroup moves at the same location when they reach an operation type.

Example of use case:

* We use several operation types for Picking Transfers (Pick Shelves, Pick Reserve)
* We use several operation types for Packing (Pack, Pack Post)
* The "Checkout Synchronization" flag is activated on the 2 Packing operation types
* The Packing can occur in several areas (Packing/PACK1, Packing/PACK2, Packing/PACK3)

Based on these assumptions, we have these pickings, all for the same order:

* SHELF001, pick 2 goods in the shelves
* RESER001, pick 2 goods in the reserve
* PACK001, pack 3 of the goods (2 from SHELF001, 1 from RESER001)
* PPOST001, pack the remaining good of the reserve

The operator uses the button "Checkout Sync" on a pick transfer,
for instance SHELF001, and select Packing/PACK2 as destination.
We want all the lines that reach PACK001 to be updated with:

* The move lines (except the one that goes to PPOST001) must now reach
  Packing/PACK2
* The same moves' destination must reach Packing/PACK2 as well:
  we constrain the user to move the goods there as we want to move
  all the goods at the same place for the packing.

Selection_848

Works best when used with ``stock_available_to_promise_release``:

* When using the mentioned module, we assume that we release moves (which
  creates the whole chain of moves) only when we know that we have the
  quantities in stock (otherwise the module splits the delivery). Any exception
  (which is now rare when you release this ways) that would lead to a backorder
  is handled by canceling the remaining line.
* When the module is not used, then the destination of the backorders may be
  changed at the same time

Compatible with ``stock_dynamic_routing``.

### CODE REVIEW

File: stock_checkout_sync/models/stock_move.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+            if not dest_pickings:
+                continue
+
+            moves = (move | move.common_dest_move_ids).filtered(
+                lambda move: move.state not in ("done", "cancel")
+            )
+            for dest_picking in dest_pickings:
+                if dest_picking not in selected_pickings:
+                    selected_pickings[dest_picking] = self.env["stock.move"].browse()
+                selected_pickings[dest_picking] |= moves
```
Comment (simahawk) [inline stock_checkout_sync/models/stock_move.py:50]: shall we use ids here instead of recordsets?

**Review by simahawk — APPROVED** [review]

File: stock_checkout_sync/models/stock_move.py (hunk 1 above)
Comment (guewen) [reply stock_checkout_sync/models/stock_move.py:50]: I don't really expect to have thousands of ids to make a difference here. It's used on one order, so a dozen of moves, maybe a hundred for a very large order.

File: stock_checkout_sync/models/stock_picking.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    def _compute_can_sync_to_checkout(self):
+        for picking in self:
+            picking.can_sync_to_checkout = False
+            for dest_move in picking.mapped("move_lines.move_dest_ids"):
+                if dest_move.picking_type_id.checkout_sync:
+                    picking.can_sync_to_checkout = True
+                    break
+
+    def open_checkout_sync_wizard(self, done_picking_types=None):
+        return self.env["stock.move.checkout.sync"]._open(self)
```
Comment (grindtildeath) [inline stock_checkout_sync/models/stock_picking.py:25]: Isn't it simpler to rely on `env.context.active_id` in `default_get` of the wizard and return an action here?

File: stock_checkout_sync/models/stock_picking.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    )
+
+    @api.depends("move_lines.move_dest_ids")
+    def _compute_can_sync_to_checkout(self):
+        for picking in self:
+            picking.can_sync_to_checkout = False
+            for dest_move in picking.mapped("move_lines.move_dest_ids"):
+                if dest_move.picking_type_id.checkout_sync:
+                    picking.can_sync_to_checkout = True
+                    break
```
Comment (grindtildeath) [inline stock_checkout_sync/models/stock_picking.py:22]:
```suggestion
            if any(
                dest_move.picking_type_id.checkout_sync for dest_move in picking.mapped(
                    "move_lines.move_dest_ids"
                )
            ):
                    picking.can_sync_to_checkout = True
```
Not sure if this makes the condition more "readable"

File: stock_move_common_dest/models/stock_move.py [hunk 4]
```diff
@@ -47,3 +52,16 @@ def _compute_common_dest_move_ids(self):
                 move.common_dest_move_ids = [(6, 0, common_move_ids)]
             else:
                 move.common_dest_move_ids = [(5, 0, 0)]
+
+    def _search_compute_dest_move_ids(self, operator, value):
```
Comment (grindtildeath) [inline stock_move_common_dest/models/stock_move.py]:
```suggestion
    def _search_common_dest_move_ids(self, operator, value):
```

File: stock_move_common_dest/models/stock_move.py [hunk 5]
```diff
@@ -10,17 +10,12 @@ class StockMove(models.Model):
     common_dest_move_ids = fields.Many2many(
         "stock.move",
         compute="_compute_common_dest_move_ids",
+        search="_search_compute_dest_move_ids",
```
Comment (grindtildeath) [inline stock_move_common_dest/models/stock_move.py]:
```suggestion
        search="_search_common_dest_move_ids",
```

File: stock_checkout_sync/models/stock_move.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+                and m in self.move_dest_ids
+            )
+            and m.location_id != location
+        ).write({"location_id": location.id})
+        lines = moves.mapped("move_line_ids")
+        lines.filtered(
+            lambda l: l.location_dest_id != location and l.state != "done"
+        ).write({"location_dest_id": location.id})
+
+    def _moves_to_sync_checkout(self):
```
Comment (grindtildeath) [inline stock_checkout_sync/models/stock_move.py:33]:
```suggestion
    def _select_pickings_to_sync_checkout(self):
```
I find it odd to return picking with the actual function name

**Review by grindtildeath — COMMENTED** [review]
Looks good overall, just some functions names to fix and questions.

File: stock_checkout_sync/models/stock_picking.py (hunk 2 above)
Comment (guewen) [reply stock_checkout_sync/models/stock_picking.py:25]: The wizard reopens itself, that's why the action is returned by a method `_open` in the wizard class. Then we need a button on the picking, and we can reuse the `_open` method of the wizard as it exists.

File: stock_checkout_sync/models/stock_move.py (hunk 6 above)
Comment (guewen) [reply stock_checkout_sync/models/stock_move.py:33]: It does return moves, but grouped by pickings in a dictionary.

**Comment by simahawk** [comment]
@guewen had to solve a conflict to make pending merge aggregation work. Fill free to rebase and trash https://github.com/OCA/stock-logistics-warehouse/pull/907/commits/314f5fba683e9147005635b5cf8adb72a4e437d5

File: stock_checkout_sync/models/stock_move.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        # destination move can change, so handle this case too. (there is a
+        # glue module stock_dynamic_routing_checkout_sync).
+        moves_to_update = self.filtered(lambda m: m.location_dest_id != location)
+        moves_to_update.write({"location_dest_id": location.id})
+        # Sync the source of the destination move too, if it's still waiting.
+        moves_to_update.move_dest_ids.filtered(
+            lambda m: (
+                m.state == "waiting"
+                or m.state == "assigned"
+                and m in self.move_dest_ids
```
Comment (sebalix) [inline stock_checkout_sync/models/stock_move.py]: I'm not sure to understand this line, `m` is supposed to be part of `moves_to_update.move_dest_ids`, and `moves_to_update` is already a subset of `self` so it's always True, no?

Comment (guewen) [reply stock_checkout_sync/models/stock_move.py]: You are right, thanks

Comment (guewen) [inline stock_checkout_sync/models/stock_move.py]:
```suggestion
```

File: stock_dynamic_routing_checkout_sync/__manifest__.py [hunk 8]
```diff
@@ -0,0 +1,16 @@
+# Copyright 2019 Camptocamp (https://www.camptocamp.com)
+{
+    "name": "Stock Routing Operations - Checkout Sync",
+    "summary": "Glue module",
+    "author": "Camptocamp, Odoo Community Association (OCA)",
+    "website": "https://github.com/OCA/stock-logistics-warehouse",
+    "category": "Warehouse Management",
+    "version": "13.0.1.0.0",
+    "license": "AGPL-3",
+    "depends": ["stock_dynamic_routing", "stock_checkout_sync"],
```
Comment (sebalix) [inline stock_dynamic_routing_checkout_sync/__manifest__.py:10]: Need to add `OCA/wms` in `oca_dependencies.txt` to get `stock_dynamic_routing`?

Comment (guewen) [reply stock_dynamic_routing_checkout_sync/__manifest__.py:10]: added, thanks!

**Comment by sebalix** [comment]
The test which fails in `stock_available` is not related to this PR, I think the variants created in the failing test ( https://github.com/OCA/stock-logistics-warehouse/blob/13.0/stock_available/tests/test_stock_available.py#L31-L55 ) should have different attributes to not trigger the constraint on `flush` call.

**Comment by simahawk** [comment]
@sebalix @guewen anything we can do to unlock this?

**Comment by guewen** [comment]
I fixed the error with `stock_available`, but we have another unrelated error:

```
Traceback (most recent call last):
File "/home/travis/build/OCA/stock-logistics-warehouse/stock_vertical_lift/tests/test_inventory.py", line 38, in test_inventory_count_ops
operation = self._open_screen("inventory")
File "/home/travis/build/OCA/stock-logistics-warehouse/stock_vertical_lift/tests/common.py", line 68, in _open_screen
getattr(shuttle or self.shuttle, "switch_{}".format(mode))()
File "/home/travis/build/OCA/stock-logistics-warehouse/stock_vertical_lift/models/vertical_lift_shuttle.py", line 210, in switch_inventory
return self.action_open_screen()
File "/home/travis/build/OCA/stock-logistics-warehouse/stock_vertical_lift/models/vertical_lift_shuttle.py", line 155, in action_open_screen
operation.on_screen_open()
File "/home/travis/build/OCA/stock-logistics-warehouse/stock_vertical_lift/models/vertical_lift_operation_base.py", line 175, in on_screen_open
self.reset_steps()
File "/home/travis/build/OCA/stock-logistics-warehouse/stock_vertical_lift/models/vertical_lift_operation_inventory.py", line 196, in reset_steps
super().reset_steps()
File "/home/travis/build/OCA/stock-logistics-warehouse/stock_vertical_lift/models/vertical_lift_operation_base.py", line 167, in reset_steps
self.next_step()
File "/home/travis/build/OCA/stock-logistics-warehouse/stock_vertical_lift/models/vertical_lift_operation_base.py", line 147, in next_step
if not transition.action or transition.action(self):
File "/home/travis/build/OCA/stock-logistics-warehouse/stock_vertical_lift/models/vertical_lift_operation_inventory.py", line 36, in <lambda>
lambda self: self.select_next_inventory_line(),
File "/home/travis/build/OCA/stock-logistics-warehouse/stock_vertical_lift/models/vertical_lift_operation_inventory.py", line 242, in select_next_inventory_line
self.fetch_tray()
File "/home/travis/build/OCA/stock-logistics-warehouse/stock_vertical_lift/models/vertical_lift_operation_inventory.py", line 231, in fetch_tray
location.fetch_vertical_lift_tray()
File "/home/travis/build/OCA/stock-logistics-warehouse/stock_vertical_lift/models/stock_location.py", line 140, in fetch_vertical_lift_tray
tray.fetch_vertical_lift_tray(cell_location=self)
File "/home/travis/build/OCA/stock-logistics-warehouse/stock_vertical_lift/models/stock_location.py", line 142, in fetch_vertical_lift_tray
self._hardware_vertical_lift_fetch_tray(cell_location=cell_location)
File "/home/travis/build/OCA/stock-logistics-warehouse/stock_vertical_lift/models/stock_location.py", line 67, in _hardware_vertical_lift_fetch_tray
payload = self._hardware_vertical_lift_fetch_tray_payload(cell_location)
File "/home/travis/build/OCA/stock-logistics-warehouse/stock_vertical_lift/models/stock_location.py", line 118, in _hardware_vertical_lift_fetch_tray_payload
raise NotImplementedError()
NotImplementedError
```

This one happens because https://github.com/OCA/stock-logistics-warehouse/blob/c5aa718b4cc8ba16dfc5b8df5885d170f8587138/stock_vertical_lift/models/stock_location.py#L35-L39

is not computed.

It fails after: https://github.com/odoo/odoo/commit/3a6ac95ddf6a8719aa714a83fc339ba838c3811d

**Comment by guewen** [comment]
Tests are now successful after rebase (fixed by #1035).

**Review by sebalix — APPROVED** [review]

**Comment by guewen** [comment]
Note: do not merge, will be moved to OCA/wms as asked by @jgrandguillaume

**Comment by simahawk** [comment]
@guewen shall we close this in favor of https://github.com/OCA/wms/pull/115 ?

**Pull request closed without merging**